File: 1748-best-team-with-no-conflicts/best-team-with-no-conflicts.py
import sys

class Solution:
    def bestTeamScore(self, scores: List[int], ages: List[int]) -> int:
        # 1. Increase recursion limit for N = 1000
        # sys.setrecursionlimit(2000)
        
        # 2. Sort to handle age constraints
        players = sorted(zip(ages, scores))
        n = len(players)
        
        # 3. Initialize manual memo table with -1 (meaning "not yet calculated")
        # Shape: [n][n + 1] to account for prev_idx = -1
        memo = [[-1] * (n + 1) for _ in range(n)]

        def solve(idx, prev_idx):
            if idx == n:
                return 0
            
            # Map prev_idx (-1 to n-1) to (0 to n) for array indexing
            memo_col = prev_idx + 1
            
            if memo[idx][memo_col] != -1:
                return memo[idx][memo_col]
            
            # Option 1: Skip current player
            res = solve(idx + 1, prev_idx)
            
            # Option 2: Pick current player (if score is non-decreasing)
            curr_score = players[idx][1]
            if prev_idx == -1 or curr_score >= players[prev_idx][1]:
                res = max(res, curr_score + solve(idx + 1, idx))
            
            # Save and return
            memo[idx][memo_col] = res
            return res

        return solve(0, -1)

File: 1748-best-team-with-no-conflicts/best-team-with-no-conflicts.java
import java.util.*;

class Solution {
    int[][] players;
    Integer[][] memo;

    public int bestTeamScore(int[] scores, int[] ages) {
        int n = scores.length;
        players = new int[n][2];
        
        // 1. Pair age and score
        for (int i = 0; i < n; i++) {
            players[i][0] = ages[i];
            players[i][1] = scores[i];
        }

        // 2. Sort by Age (primary) and Score (secondary)
        Arrays.sort(players, (a, b) -> a[0] == b[0] ? a[1] - b[1] : a[0] - b[0]);

        // 3. Initialize memoization table
        // We use n + 1 for the second dimension to handle the -1 index case
        memo = new Integer[n][n + 1];

        return solve(0, -1, n);
    }

    private int solve(int idx, int prevIdx, int n) {
        if (idx == n) return 0;

        // prevIdx + 1 maps -1 to 0, 0 to 1, etc., for the memo table
        if (memo[idx][prevIdx + 1] != null) {
            return memo[idx][prevIdx + 1];
        }

        // Option 1: Skip the current player
        int res = solve(idx + 1, prevIdx, n);

        // Option 2: Pick the current player (if no conflict)
        if (prevIdx == -1 || players[idx][1] >= players[prevIdx][1]) {
            res = Math.max(res, players[idx][1] + solve(idx + 1, idx, n));
        }

        return memo[idx][prevIdx + 1] = res;
    }
}